icu4c/source/i18n/number_longnames.cpp

```diff
@@ -295,17 +440,47 @@ void LongNameHandler::forCompoundUnit(const Locale &loc, const MeasureUnit &unit
         status = U_INTERNAL_PROGRAM_ERROR;
         return;
     }
+
+    StringPiece primaryCase, secondaryCase;
+    CharString primaryCase_, secondaryCase_;
+    if (!unitDisplayCase.empty()) {
+        U_ASSERT(U_SUCCESS(status));
+        UnicodeString uVal0, uVal1;
+        getDeriveComponentRule(loc, "case", "per", uVal0, uVal1, status);
+        if (U_SUCCESS(status)) {
+            if (uVal0.compare(UnicodeString(u"compound")) == 0) {
+                primaryCase = unitDisplayCase;
+            } else {
+                primaryCase_.appendInvariantChars(uVal0, status);
+                primaryCase = primaryCase_.toStringPiece();
+            }
+            if (uVal1.compare(UnicodeString(u"compound")) == 0) {
+                secondaryCase = unitDisplayCase;
+            } else {
+                secondaryCase_.appendInvariantChars(uVal1, status);
+                secondaryCase = secondaryCase_.toStringPiece();
+            }
+        } else {
+            // Ignore failures in production code, assert-fail in debug mode.
+            U_ASSERT(false);
+            status = U_ZERO_ERROR;
+        }
+    }
```

Comment:
nit/ many nested if statements

Reply:
Ooh, I like this idea:
I could refactor "getDeriveComponentRule" in a way that produces primaryCase and secondaryCase instead of producing uVal0 and uVal1.

Reply:
Main problem: needing to take care of storing the converted strings means I would have to pass in primaryCaseMem, secondaryCaseMem, primaryCase, and secondaryCase. Might actually make sense to make a little class out of this.

Reply:
Conceptually:

    DerivedComponents derivedCases(loc, "case", "per");
    primaryCase = derivedCases.value0;
    secondaryCase = derivedCases.value1;

Means I can reuse it for "times", "power" and "prefix, as well as using it for both "case" and "plural"!

Reply:
Done! PTAL.

I'm not doing the same thing for deriveCompound because it does not look like it has much benefit: all it could do is in any case return a 0 or 1. (Could be reconsidered once I'm using `getDeriveCompoundRule` more.)